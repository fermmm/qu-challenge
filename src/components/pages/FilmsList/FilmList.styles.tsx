// tslint:disable: typedef
import styled from 'styled-components';

export const FilmListContainer = styled('div')`
   display: flex;
   align-items: center;
   justify-content: center;
   flex-direction: column;
   padding: 30px;
   width: 100%;
   max-width: 650px;
`;

export const SortControlsContainer = styled('div')`
   display: flex;
   flex-direction: row;
   margin-bottom: 20px;
`;

export const SortButton = styled('button')`
   margin: 0 6px;
   cursor: pointer;
`;
