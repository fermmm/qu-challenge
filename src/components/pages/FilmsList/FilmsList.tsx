import React, { FC, memo, useEffect, useState } from 'react';
import { hot } from 'react-hot-loader';
import { firstBy } from 'thenby';
import { Film } from '../../../api/tools/typings';
import { Column } from '../../common/Column/Column';
import { Row } from '../../common/Row/Row';
import FilmItem from './FilmItem/FilmItem';

interface PropsFilmsList {
   films: Film[];
}

export const FilmsList: FC<PropsFilmsList> = memo(({ films }) => {
   const [filmsSorted, setFilmsSorted] = useState(films);
   const [sortBy, setSortBy] = useState<keyof Film>();

   useEffect(() => {
      setFilmsSorted(sortList(films));
   }, [films, sortBy]);

   /**
    * Here I could implement my own sort function but I decided to use
    * the "thenby" module because in my opinion it's a better demonstration
    * since this module is very powerful to implement very sophisticated
    * composed sorting in no time:
    * https://github.com/Teun/thenBy.js/tree/master
    */
   const sortList = (filmsUnsorted: Film[]) =>
      filmsUnsorted && [...filmsUnsorted].sort(firstBy(f => f[sortBy]));

   return (
      <Column>
         <Row>
            Sort by:
            <button onClick={() => setSortBy('title')}>Title</button>
            <button onClick={() => setSortBy('release_date')}>Release Date</button>
         </Row>
         {filmsSorted &&
            filmsSorted.map(filmData => <FilmItem key={filmData.episode_id} filmData={filmData} />)}
      </Column>
   );
});

export default hot(module)(FilmsList);
