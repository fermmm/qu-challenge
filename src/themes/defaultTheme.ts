import { Theme } from './typings/Theme';

export const defaultTheme: Theme = {
   colors: {
      background1: '#FFFFFF',
      text1: '#A8A8A8',
      scrollbar: '#FFFFFF',
      scrollbarThumb: '#565656',
   },
   shadows: {
      elevation1: '1px 1px 6px 0px #00000012',
   },
   fonts: {
      font1: '"Montserrat", sans-serif',
      size1: '22px',
      weight1: '600',
      letterSpacing1: '-0.26px',
      lineHeight1: '21px',
   },
   borders: {
      borderRadius1: '5px',
   },
};
