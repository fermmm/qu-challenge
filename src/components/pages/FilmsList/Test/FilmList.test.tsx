import { mount } from 'enzyme';
import React from 'react';
import { ThemeProvider } from 'styled-components';
import { defaultTheme } from '../../../../themes/defaultTheme';
import { FilmsList } from '../FilmsList';
// @ts-ignore
import dummyData from './dummyData.json';

describe('<FilmList />', () => {
   test('Renders correctly', () => {
      const component = mount(
         <ThemeProvider theme={defaultTheme}>
            <FilmsList films={dummyData} />
         </ThemeProvider>
      );
      expect(component.html()).toMatchSnapshot();
      component.unmount();
   });

   test('Sorting works', () => {
      const component = mount(
         <ThemeProvider theme={defaultTheme}>
            <FilmsList films={dummyData} />
         </ThemeProvider>
      );
      component.find('button').at(0).simulate('click');
      expect(component.html()).toMatchSnapshot();
      component.find('button').at(1).simulate('click');
      expect(component.html()).toMatchSnapshot();
      component.unmount();
   });
});
