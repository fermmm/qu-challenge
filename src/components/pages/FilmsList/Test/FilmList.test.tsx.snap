// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilmList /> Renders correctly 1`] = `"<div class=\\"FilmListstyles__FilmListContainer-sc-1g4vwsu-0 bKKCvi\\"><div class=\\"FilmListstyles__SortControlsContainer-sc-1g4vwsu-1 hpWsaf\\">Sort by:<button class=\\"FilmListstyles__SortButton-sc-1g4vwsu-2 hVlahb\\">Title</button><button class=\\"FilmListstyles__SortButton-sc-1g4vwsu-2 hVlahb\\">Release Date</button></div><div class=\\"FilmItemstyles__FilmItemContainer-sc-14hvbk9-0 eSGjBp\\"><div class=\\"Row-sc-10eqdlj-0 hATPeQ\\">A New Hope</div></div><div class=\\"FilmItemstyles__FilmItemContainer-sc-14hvbk9-0 eSGjBp\\"><div class=\\"Row-sc-10eqdlj-0 hATPeQ\\">The Empire Strikes Back</div></div><div class=\\"FilmItemstyles__FilmItemContainer-sc-14hvbk9-0 eSGjBp\\"><div class=\\"Row-sc-10eqdlj-0 hATPeQ\\">Return of the Jedi</div></div><div class=\\"FilmItemstyles__FilmItemContainer-sc-14hvbk9-0 eSGjBp\\"><div class=\\"Row-sc-10eqdlj-0 hATPeQ\\">The Phantom Menace</div></div><div class=\\"FilmItemstyles__FilmItemContainer-sc-14hvbk9-0 eSGjBp\\"><div class=\\"Row-sc-10eqdlj-0 hATPeQ\\">Attack of the Clones</div></div><div class=\\"FilmItemstyles__FilmItemContainer-sc-14hvbk9-0 eSGjBp\\"><div class=\\"Row-sc-10eqdlj-0 hATPeQ\\">Revenge of the Sith</div></div></div>"`;

exports[`<FilmList /> Sorting works 1`] = `"<div class=\\"FilmListstyles__FilmListContainer-sc-1g4vwsu-0 bKKCvi\\"><div class=\\"FilmListstyles__SortControlsContainer-sc-1g4vwsu-1 hpWsaf\\">Sort by:<button class=\\"FilmListstyles__SortButton-sc-1g4vwsu-2 hVlahb\\">Title</button><button class=\\"FilmListstyles__SortButton-sc-1g4vwsu-2 hVlahb\\">Release Date</button></div><div class=\\"FilmItemstyles__FilmItemContainer-sc-14hvbk9-0 eSGjBp\\"><div class=\\"Row-sc-10eqdlj-0 hATPeQ\\">A New Hope</div></div><div class=\\"FilmItemstyles__FilmItemContainer-sc-14hvbk9-0 eSGjBp\\"><div class=\\"Row-sc-10eqdlj-0 hATPeQ\\">Attack of the Clones</div></div><div class=\\"FilmItemstyles__FilmItemContainer-sc-14hvbk9-0 eSGjBp\\"><div class=\\"Row-sc-10eqdlj-0 hATPeQ\\">Return of the Jedi</div></div><div class=\\"FilmItemstyles__FilmItemContainer-sc-14hvbk9-0 eSGjBp\\"><div class=\\"Row-sc-10eqdlj-0 hATPeQ\\">Revenge of the Sith</div></div><div class=\\"FilmItemstyles__FilmItemContainer-sc-14hvbk9-0 eSGjBp\\"><div class=\\"Row-sc-10eqdlj-0 hATPeQ\\">The Empire Strikes Back</div></div><div class=\\"FilmItemstyles__FilmItemContainer-sc-14hvbk9-0 eSGjBp\\"><div class=\\"Row-sc-10eqdlj-0 hATPeQ\\">The Phantom Menace</div></div></div>"`;

exports[`<FilmList /> Sorting works 2`] = `"<div class=\\"FilmListstyles__FilmListContainer-sc-1g4vwsu-0 bKKCvi\\"><div class=\\"FilmListstyles__SortControlsContainer-sc-1g4vwsu-1 hpWsaf\\">Sort by:<button class=\\"FilmListstyles__SortButton-sc-1g4vwsu-2 hVlahb\\">Title</button><button class=\\"FilmListstyles__SortButton-sc-1g4vwsu-2 hVlahb\\">Release Date</button></div><div class=\\"FilmItemstyles__FilmItemContainer-sc-14hvbk9-0 eSGjBp\\"><div class=\\"Row-sc-10eqdlj-0 hATPeQ\\">A New Hope</div></div><div class=\\"FilmItemstyles__FilmItemContainer-sc-14hvbk9-0 eSGjBp\\"><div class=\\"Row-sc-10eqdlj-0 hATPeQ\\">The Empire Strikes Back</div></div><div class=\\"FilmItemstyles__FilmItemContainer-sc-14hvbk9-0 eSGjBp\\"><div class=\\"Row-sc-10eqdlj-0 hATPeQ\\">Return of the Jedi</div></div><div class=\\"FilmItemstyles__FilmItemContainer-sc-14hvbk9-0 eSGjBp\\"><div class=\\"Row-sc-10eqdlj-0 hATPeQ\\">The Phantom Menace</div></div><div class=\\"FilmItemstyles__FilmItemContainer-sc-14hvbk9-0 eSGjBp\\"><div class=\\"Row-sc-10eqdlj-0 hATPeQ\\">Attack of the Clones</div></div><div class=\\"FilmItemstyles__FilmItemContainer-sc-14hvbk9-0 eSGjBp\\"><div class=\\"Row-sc-10eqdlj-0 hATPeQ\\">Revenge of the Sith</div></div></div>"`;
